### PR TITLE
Add autest to cover updates to cache with alternates

### PR DESCRIPTION
**Pull request by cmcfarlen**
This autest demonstrates an issue with cache alternates and updating from a response with a content-length to one with content-length 0.

When an alternate is replaced, and previously had a content-length, there is a bug where then old content is preserved and only this headers are changed (and the new 0-length content is not "written").  When this half-updated asset is served from cache, the content length is mismatched, delivered via chunked encoding and the old contents are shipped to the client.

### CODE REVIEW

**Comment by elsloo** [comment]
When digging into this issue, it appears to stem from setting a CacheVC instance variable called `update_len` to the value of the existing cached object during the call to [`Cache::open_write()`](https://github.com/apache/trafficserver/blob/master/iocore/cache/CacheWrite.cc#L1818) which occurs prior to the request being sent upstream on cache miss.

Later in the transaction, when a response with a `Content-Length` of zero is received and being handled, [the length of the resulting object is set incorrectly to the prior length using `update_len`](https://github.com/apache/trafficserver/blob/master/iocore/cache/Cache.cc#L448). The first request to go through this transition will be served as expected, while any subsequent request will hang, waiting on the response because of ATS unexpectedly switching to `Transfer-Encoding: chunked`. The [length mismatch within ATS arises](https://github.com/apache/trafficserver/blob/master/proxy/http/HttpTransact.cc#L6799) from the newly cached headers that contain `Content-Length: 0` and the length of the fragment on disk which was set incorrectly using `update_len`. Clients requesting objects in this state will eventually abort.

Because we are already reading the `Content-Length` header in `CacheVC::set_http_info()` and specifically handling the zero length case, we could easily handle the issue there. However, there could be other issues that arise from a mismatch in this scenario where the `Content-Length` is non-zero but greater than or less than the value set on `update_len` so we should consider the spirit of the intent of the alternate update process that sets `update_len` during `Cache::open_write()` while mitigating the underlying cause.

**Review by bneradt — APPROVED** [review]
The test looks good. Thanks for adding a regression test for this.

**Comment by ywkaras** [comment]
[approve ci autest]

**Comment by ywkaras** [comment]
Looks like this is causing the  alternate-caching Au test to fail.
```
Running Test all_headers:... Passed
Running Test alternate-caching:F Failed
Running Test background_fill:... Passed
```

**Comment by cmcfarlen** [comment]
Yes, this PR is just an autest that demonstrates the bug with alternate cache updates so that people could reason about it.  A fix is forthcoming and then this autest will pass.

**Comment by bneradt** [comment]
[approve ci fedora]

File: iocore/cache/Cache.cc [hunk 1]
```diff
@@ -454,6 +454,7 @@ CacheVC::set_http_info(CacheHTTPInfo *ainfo)
   MIMEField *field = ainfo->m_alt->m_response_hdr.field_find(MIME_FIELD_CONTENT_LENGTH, MIME_LEN_CONTENT_LENGTH);
   if (field && !field->value_get_int64()) {
     f.allow_empty_doc = 1;
+    ainfo->object_size_set(0);
```
Comment (bneradt) [inline iocore/cache/Cache.cc:457]: Probably good to comment why this line is here.

**Review by bneradt — CHANGES_REQUESTED** [review]

**Comment by bneradt** [comment]
Looks good.

**Review by bneradt — APPROVED** [review]
Looks good to me.

**Comment by elsloo** [comment]
Merging this PR on behalf of @cmcfarlen due to the approval from @bneradt and my own knowledge of how this works.

**Comment by zwoop** [comment]
Cherry-picked to v9.2.x

**Comment by masaori335** [comment]
We faced another symptom made by this bug. If the half-updated object is big enough (> 1GB) and it's on the RAM Cache, you might see the below crash made by a huge loop on cache read.

```
[ 00 ] libpthread-2.17.so  waitpid                                                           
[ 01 ] traffic_server      crash_logger_invoke(int, siginfo_t*, void*)                       ( Crash.cc:168 )
[ 02 ] libpthread-2.17.so                                                                    
[ 03 ] libtscore.so.9.0.3  freelist_new(_InkFreeList*)                                       ( ink_queue.cc:239 )
[ 04 ] libtscore.so.9.0.3  ink_freelist_new                                                  ( ink_queue.cc:187 )
[ 05 ] traffic_server      IOBufferBlock::alloc(long)                                        ( Allocator.h:131 )
[ 06 ] traffic_server      MIOBuffer::write(void const*, long)                               ( P_IOBuffer.h:839 )
[ 07 ] traffic_server      ChunkedHandler::generate_chunked_content()                        ( HttpTunnel.cc:371 )
[ 08 ] traffic_server      HttpTunnel::producer_handler_dechunked(int, HttpTunnelProducer*)  ( HttpTunnel.cc:1019 )
[ 09 ] traffic_server      HttpTunnel::producer_handler(int, HttpTunnelProducer*)            ( HttpTunnel.cc:1116 )
[ 10 ] traffic_server      HttpTunnel::main_handler(int, void*)                              ( HttpTunnel.cc:1640 )
[ 11 ] traffic_server      CacheVC::calluser(int)                                            ( eventsystem/I_Continuation.h:219 )
[ 12 ] traffic_server      CacheVC::openReadMain(int, Event*)                                ( CacheRead.cc:782 )
[ 13 ] traffic_server      CacheVC::openReadReadDone(int, Event*)                            ( CacheRead.cc:639 )
[ 14 ] traffic_server      CacheVC::openReadMain(int, Event*)                                ( eventsystem/I_Continuation.h:219 )
[ 15 ] traffic_server      CacheVC::openReadReadDone(int, Event*)                            ( CacheRead.cc:639 )
[ 16 ] traffic_server      CacheVC::openReadMain(int, Event*)                                ( eventsystem/I_Continuation.h:219 )
[ 17 ] traffic_server      CacheVC::openReadReadDone(int, Event*)                            ( CacheRead.cc:639 )
[ 18 ] traffic_server      CacheVC::openReadMain(int, Event*)                                ( eventsystem/I_Continuation.h:219 )
[ 19 ] traffic_server      CacheVC::openReadReadDone(int, Event*)                            ( CacheRead.cc:639 )
[ 20 ] traffic_server      CacheVC::openReadMain(int, Event*)                                ( eventsystem/I_Continuation.h:219 )
[ 21 ] traffic_server      CacheVC::openReadReadDone(int, Event*)                            ( CacheRead.cc:639 )
[ 22 ] traffic_server      CacheVC::openReadMain(int, Event*)                                ( eventsystem/I_Continuation.h:219 )
[ 23 ] traffic_server      CacheVC::openReadReadDone(int, Event*)                            ( CacheRead.cc:639 )
[ 24 ] traffic_server      CacheVC::openReadMain(int, Event*)                                ( eventsystem/I_Continuation.h:219 )
[ 25 ] traffic_server      CacheVC::openReadReadDone(int, Event*)                            ( CacheRead.cc:639 )
[ 26 ] traffic_server      CacheVC::openReadMain(int, Event*)                                ( eventsystem/I_Continuation.h:219 )
[ 27 ] traffic_server      CacheVC::openReadReadDone(int, Event*)                            ( CacheRead.cc:639 )
[ 28 ] traffic_server      CacheVC::openReadMain(int, Event*)                                ( eventsystem/I_Continuation.h:219 )
[ 29 ] traffic_server      CacheVC::openReadReadDone(int, Event*)                            ( CacheRead.cc:639 )
[ 30 ] traffic_server      CacheVC::openReadMain(int, Event*)                                ( eventsystem/I_Continuation.h:219 )
[ 31 ] traffic_server      CacheVC::openReadReadDone(int, Event*)                            ( CacheRead.cc:639 )
[ 32 ] traffic_server      CacheVC::openReadMain(int, Event*)                                ( eventsystem/I_Continuation.h:219 )
[ 33 ] traffic_server      CacheVC::openReadReadDone(int, Event*)                            ( CacheRead.cc:639 )
[ 34 ] traffic_server      CacheVC::openReadMain(int, Event*)                                ( eventsystem/I_Continuation.h:219 )
[ 35 ] traffic_server      CacheVC::openReadReadDone(int, Event*)                            ( CacheRead.cc:639 )
[ 36 ] traffic_server      CacheVC::openReadMain(int, Event*)                                ( eventsystem/I_Continuation.h:219 )
[ 37 ] traffic_server      CacheVC::openReadReadDone(int, Event*)                            ( CacheRead.cc:639 )
[ 38 ] traffic_server      CacheVC::openReadMain(int, Event*)                                ( eventsystem/I_Continuation.h:219 )
[ 39 ] traffic_server      CacheVC::openReadReadDone(int, Event*)                            ( CacheRead.cc:639 )
[ 40 ] traffic_server      CacheVC::openReadMain(int, Event*)                                ( eventsystem/I_Continuation.h:219 )
[ 41 ] traffic_server      CacheVC::openReadReadDone(int, Event*)                            ( CacheRead.cc:639 )
[ 42 ] traffic_server      CacheVC::openReadMain(int, Event*)                                ( eventsystem/I_Continuation.h:219 )
[ 43 ] traffic_server      CacheVC::openReadReadDone(int, Event*)                            ( CacheRead.cc:639 )
[ 44 ] traffic_server      CacheVC::openReadMain(int, Event*)                                ( eventsystem/I_Continuation.h:219 )
[ 45 ] traffic_server      CacheVC::openReadReadDone(int, Event*)                            ( CacheRead.cc:639 )
[ 46 ] traffic_server      CacheVC::openReadMain(int, Event*)                                ( eventsystem/I_Continuation.h:219 )
[ 47 ] traffic_server      CacheVC::openReadReadDone(int, Event*)                            ( CacheRead.cc:639 )
[ 48 ] traffic_server      CacheVC::openReadMain(int, Event*)                                ( eventsystem/I_Continuation.h:219 )
[ 49 ] traffic_server      CacheVC::openReadReadDone(int, Event*)                            ( CacheRead.cc:639 )
[ 50 ] traffic_server      CacheVC::openReadMain(int, Event*)                                ( eventsystem/I_Continuation.h:219 )
[ 51 ] traffic_server      CacheVC::openReadReadDone(int, Event*)                            ( CacheRead.cc:639 )
[ 52 ] traffic_server      CacheVC::openReadMain(int, Event*)                                ( eventsystem/I_Continuation.h:219 )
[ 53 ] traffic_server      CacheVC::openReadReadDone(int, Event*)                            ( CacheRead.cc:639 )
[ 54 ] traffic_server      CacheVC::openReadMain(int, Event*)                                ( eventsystem/I_Continuation.h:219 )
[ 55 ] traffic_server      CacheVC::openReadReadDone(int, Event*)                            ( CacheRead.cc:639 )
[ 56 ] traffic_server      CacheVC::openReadMain(int, Event*)                                ( eventsystem/I_Continuation.h:219 )
[ 57 ] traffic_server      CacheVC::openReadReadDone(int, Event*)                            ( CacheRead.cc:639 )
[ 58 ] traffic_server      CacheVC::openReadMain(int, Event*)                                ( eventsystem/I_Continuation.h:219 )
[ 59 ] traffic_server      CacheVC::openReadReadDone(int, Event*)                            ( CacheRead.cc:639 )
[ 60 ] traffic_server      CacheVC::openReadMain(int, Event*)                                ( eventsystem/I_Continuation.h:219 )
[ 61 ] traffic_server      CacheVC::openReadReadDone(int, Event*)                            ( CacheRead.cc:639 )
[ 62 ] traffic_server      CacheVC::openReadMain(int, Event*)                                ( eventsystem/I_Continuation.h:219 )
[ 63 ] traffic_server      CacheVC::openReadReadDone(int, Event*)                            ( CacheRead.cc:639 )
...                                                           
```
( the number of call stack is over 8000 )

**Pull request merged**